kvdb-web/src/indexed_db.rs

```diff
@@ -0,0 +1,227 @@
+		let req = target.dyn_ref::<IdbRequest>().expect_throw("Event target is IdbRequest");
+		let result = req.result().expect_throw("IdbRequest should have a result");
+		let db: &IdbDatabase = result.unchecked_ref();
+
+		for name in (0..=columns).map(store_name) {
+			let res = db.create_object_store(name.as_str());
+			if let Err(err) = res {
+				debug!("error creating object store {}: {:?}", name, err);
+			}
+		}
+	});
+
+	req.set_onupgradeneeded(Some(on_upgradeneeded.as_ref().unchecked_ref()));
+	on_upgradeneeded.forget();
+}
+
+/// Commit a transaction to the IndexedDB.
+pub fn idb_commit_transaction(
+	idb: &IdbDatabase,
+	txn: &DBTransaction,
+	columns: u32,
+) -> impl Future<Output = ()> {
+	let store_names_js = store_names_js(columns);
+
+	// Create a transaction
+	let mode = IdbTransactionMode::Readwrite;
+	let idb_txn = idb.transaction_with_str_sequence_and_mode(&store_names_js, mode)
+		.expect_throw("Failed to create an IndexedDB transaction");
+
+	// Open object stores (columns)
+	let object_stores = (0..=columns).map(|n| {
+		idb_txn.object_store(store_name(n).as_str())
+			.expect("Object stores were created in try_create_object_stores; qed")
+	}).collect::<Vec<_>>();
+
+	for op in &txn.ops {
+		match op {
+			DBOp::Insert { col, key, value } => {
+				let column = column_to_number(*col) as usize;
+
+				// Convert rust bytes to js arrays
+				// TODO (optimize): we could avoid copies
+				// with unsafe code (`Uint8Array::view`)
+				let key_js = Uint8Array::from(key.as_ref());
+				let val_js = Uint8Array::from(value.as_ref());
+
+				// Insert key/value pair into the object store
+				let res = object_stores[column].put_with_key(val_js.as_ref(), key_js.as_ref());
+				if let Err(err) = res {
+					warn!("error inserting key/values into col_{}: {:?}", column, err);
+				}
+			},
+			DBOp::Delete { col, key } => {
+				let column = column_to_number(*col) as usize;
+
+				// Convert rust bytes to js arrays
+				let key_js = Uint8Array::from(key.as_ref());
+
+				// Delete key/value pair from the object store
+				let res = object_stores[column].delete(key_js.as_ref());
+				if let Err(err) = res {
+					warn!("error deleting key from col_{}: {:?}", column, err);
+				}
+			},
+		}
+	}
+
+	let (tx, rx) = channel::oneshot::channel::<()>();
+
+	let on_complete = Closure::once(move || {
+		let _ = tx.send(());
+	});
+	idb_txn.set_oncomplete(Some(on_complete.as_ref().unchecked_ref()));
+	on_complete.forget();
+	// TODO: handle idb_txn.onerror
+
+	rx.map(|_| ())
+}
+
+
+/// Returns a cursor to a database column with the given column number.
+pub fn idb_cursor(idb: &IdbDatabase, col: u32) -> impl Stream<Item = (Vec<u8>, Vec<u8>)> {
+	// TODO: we could read all the columns in one db transaction
+	let store_name = store_name(col);
+	let store_name = store_name.as_str();
+	let txn = idb.transaction_with_str(store_name)
+		.expect_throw("Failed to create an IndexedDB transaction");
```

Comment:
Shouldn't that just be `expect` instead of `expect_throw`?
```suggestion
		.expect("Failed to create an IndexedDB transaction");
```

(same below)